pydatastructs/trees/tests/test_binary_trees.py

```diff
@@ -1,11 +1,220 @@
+    assert bl1.lowest_common_ancestor(55, 100, 2) == 90
+    assert bl1.lowest_common_ancestor(16, 80, 2) == 50
+    assert bl1.lowest_common_ancestor(30, 55, 2) == 50
+
+    assert raises(ValueError, lambda: bl1.lowest_common_ancestor(60, 200, 2))
+    assert raises(ValueError, lambda: bl1.lowest_common_ancestor(200, 60, 2))
+    assert raises(ValueError, lambda: bl1.lowest_common_ancestor(-3, 4, 2))
+
+    bl2 = BST(backend=Backend.CPP)
+    nodes = [50, 30, 90, 70, 100, 60, 80, 55, 20, 40, 15, 10, 16, 17, 18]
+    for node in nodes:
+        bl2.insert(node, node)
+    assert str(bl2) == "[(1, 50, 50, 2), (8, 30, 30, 9), (3, 90, 90, 4), (5, 70, 70, 6), (None, 100, 100, None), (7, 60, 60, None), (None, 80, 80, None), (None, 55, 55, None), (10, 20, 20, None), (None, 40, 40, None), (11, 15, 15, 12), (None, 10, 10, None), (None, 16, 16, 13), (None, 17, 17, 14), (None, 18, 18, None)]"
+
+    ##### lowest common ancestor _lca1_() tests #####
+    assert bl2.lowest_common_ancestor(80, 55, 1) == 70
+    assert bl2.lowest_common_ancestor(60, 70, 1) == 70
+    assert bl2.lowest_common_ancestor(18, 18, 1) == 18
+    assert bl2.lowest_common_ancestor(40, 90, 1) == 50
+
+    assert bl2.lowest_common_ancestor(18, 10, 1) == 15
+    assert bl2.lowest_common_ancestor(55, 100, 1) == 90
+    assert bl2.lowest_common_ancestor(16, 80, 1) == 50
+    assert bl2.lowest_common_ancestor(30, 55, 1) == 50
+
+    assert raises(ValueError, lambda: bl2.lowest_common_ancestor(60, 200, 1))
+    assert raises(ValueError, lambda: bl2.lowest_common_ancestor(200, 60, 1))
+    assert raises(ValueError, lambda: bl2.lowest_common_ancestor(-3, 4, 1))
+
+    ##### rank() tests #####
+    assert bl2.rank(18) == 5
+    assert bl2.rank(10) == 1
+    rank_list = [2, 2, 4, 4, 5, 4, 5, 3, 2, 3, 2, 1, 3, 4, 5]
+    for i,node in enumerate(nodes):
+        assert bl2.rank(node) == rank_list[i]
+    assert bl2.rank(200) is None
+
+    ##### select() tests #####
+    select_list = [10, 50, 55, 90, 100]
+    for i in range(5):
+        assert bl2.select(i+1).key == select_list[i]
+
+    b3 = BST(backend=Backend.CPP)
+    b3.insert(10, 10)
+    b3.insert(18, 18)
+    b3.insert(7, 7)
+
+    ##### upper_bound() tests #####
+    assert b3.upper_bound(9) == 10
+    assert b3.upper_bound(7) == 10
+    assert b3.upper_bound(-1) == 7
+    assert b3.upper_bound(20) is None
+
+    ##### lower_bound() tests #####
+    assert b3.lower_bound(9) == 10
+    assert b3.lower_bound(7) == 7
+    assert b3.lower_bound(-1) == 7
+    assert b3.lower_bound(20) is None
+
+
+def test_cpp_BST_speed():
+    BST = BinarySearchTree
+    import time
+    b = BST()
+    t1 = time.time()
+    for node in range(-1000,1000):
+        b.insert(node, node)
+    t2 = time.time()
+    b2 = BST(backend=Backend.CPP)
+    t3 = time.time()
+    for node in range(-1000,1000):
+        b2.insert(node, node)
+    t4 = time.time()
+    print("Time taken by Python backend: ",t2-t1,"s")
+    print("Time taken by C++ backend:    ",t4-t3,"s")
```

Comment:
See how benchmarking tests are added. Follow the same pattern.

Reply:
Done.